### ISSUE TITLE
[idea/branstorm] send to Koreader (Android share via -> Koreader)

### ISSUE THREAD

**Issue by mwoz123** [koreader/koreader#3630]
* KOReader version:
* Device:

#### Issue
Often I wish I could read article found on phone on my E-ink Koreader device.

Idea here is to use build in android, ["share via"](https://i.stack.imgur.com/4fvUy.png) option  to send it to Koreader (separate Android app/plugin) 
& on Koreader side use plugin to retrieve the data.

Android "share via" plugin will: 
- export url, 
- (maybe convert it here to PDF/HTML/EPUB?) , 
- send url/document by email
- also needed configuration window to enter email data

From Koreader side:
- write plug-in to retrieve data / check if it's possible to use Calibre plugin for that.

At glance as smartphones usually got better components than e-ink devices (and there are more libs for java than lua) I think conversion between url & desired format should be done at phone.

Guys what you think about it?

**Comment by Frenzie** [comment]
I'm inclined to think that as a user it would make more sense to set up something more generic and multi-platform like Syncthing (or Dropbox). I've been toying with the idea of writing something that'll do just that on Kobo.

(On Android that wouldn't be relevant because it already has a Syncthing app.)

Incidentally, what you describe sounds a bit like a more complicated version of:

1. Start FTP server on phone: https://f-droid.org/packages/org.primftpd/
2. Use KOReader FTP plugin to retrieve document

**Comment by mwoz123** [comment]
Point is to simplify/partially automate this process.
I dream about click "send to Koreader" and then download, conversion and sending to server will happen; 
and on Koreader click "download all the articles", and that's all;)

Maybe I'm wrong but creating basic logic for Android, doens't seem too hard..

I glanced at calibre plugin, I thought it's using email to send data (that's why I proposed email exchange above), but it seems isn't not.

Maybe then use FTP to exchange data? 
As far as I can see [Synchthing doens't allow to use FTP](https://forum.syncthing.net/t/syncthing-as-ftp-tool/8299/3), and I'd like to avoid such dependencies if possible (but would be good if could be exchange in feature e.g. FTP engine to something else e.g. this)

>Use KOReader FTP plugin to retrieve document

Is there such plugin? Or it should be created?
If yes we could extend it(make another on top of it) to pull files from specified ftp direcotry.

> to set up something more generic and multi-platform like Syncthing (or Dropbox)

hope FTP is enought generic?;) or you talking about generic phone/device app?

**Comment by KenMaltby** [comment]
There is this: https://github.com/koreader/koreader/wiki/Zsync-transport.  But I would see an advantage to being able to send from your phone wherever you are to a cloud or home server, for later retrieval by your eink device.

**Comment by Frenzie** [comment]
> Maybe then use FTP to exchange data?
As far as I can see Synchthing doens't allow to use FTP

Syncthing doesn't use FTP but I'm not really sure why that should be a problem? The whole point is that Syncthing does basically exactly 'click "download all the articles", and that's all'. ;-)

Well, except that when you say "articles" I suppose you might be talking about something more in the wallabag direction. You can use that for free here: https://framabag.org/login

For wallabag, see #3065 #2621

> Maybe I'm wrong but creating basic logic for Android, doens't seem too hard..

Making Android say "send to KOReader" in the share menu would be super simple, but what would that even mean? :-)

I suppose like @KenMaltby suggested it could somehow initiate the Zsync stuff, though that's a lot easier said than done at present. That's pretty much the only interpretation that really makes sense to me from a KOReader-exclusive perspective.

> Is there [an FTP] plugin? Or it should be created?

![screenshot_2018-01-26_18-34-03](https://user-images.githubusercontent.com/202757/35452643-7f426b4c-02c7-11e8-95ee-386488482144.png)

> hope FTP is enought generic?;) or you talking about generic phone/device app?

By generic I mean that Syncthing works on all platforms. I can toss something in a Syncthing folder on any device or computer I've got set up with Syncthing. FTP is somewhat similar but requires much more configuration. In any case, you can already use both Zsync and FTP much the way you describe.

**Comment by KenMaltby** [comment]
I think that the OP sounds like it would be something like Pocket or Wallabag, that is wanted, by the OP.  I don't use my phone that much, if I were looking up something I would be doing it on my desktop.  The other novel thing I use my eReaders for is to read books, I know kinda old fashioned.  I have a wired phone, at home also.

**Comment by mwoz123** [comment]
My analysis:

Framabag  & wallabag  
Cons:
-  seems cool but Framabag's only in French(?), 
- multiplaform (PC, phones)
- wallabag is only 14 days free, 
- can't store pages that requires login (this git page was stored only title & url), 
- requires dealing with wallabag api in lua. 
- only 1 (?) free provider - framabag?
Pros:
- Should work from both Android/Phones and PC browsers without any develpment. 

zsync: 
Cons: 
- both device have to be in the same time in the same wifi

Syncthing
Cons:
- doesn't store articles out of the box, 
- its android app is useless ("nothing to share") for us and should be replaced with custom "share via Syncthing" plugin, 
- also requires dealing with synchthing sources/api in lua
Pros:
- free, opensource 

FTP 
Cons:
- not know good free server. Most have limits. 
Pros: 
- should be easiest to write Android and Lua 

If I'd to implement that I'd go for FTP as it seems easier todo in Lua. 
Trying to keep FTP backend as interface, so it could be replaced with others later on

**Comment by Frenzie** [comment]
> wallabag is only 14 days free,

I was very confused by this statement, but it refers to the wallabag.it service.

> can't store pages that requires login (this git page was stored only title & url),

This page doesn't require login. It's misdetecting the relevant content.

> requires dealing with wallabag api in lua.

Actually you can generate an RSS feed from wallabag. ;-)

<hr>

> doesn't store articles out of the box,

Nothing that isn't wallabag or Pocket does. The whole point there is that thousands of hours have gone into wallabag for doing exactly that.

<hr>

> its android app is useless ("nothing to share") for us and should be replaced with custom "share via Syncthing" plugin,

Hm? https://github.com/syncthing/syncthing-android/pull/843

> also requires dealing with synchthing sources/api in lua

You would just run Syncthing. KOReader doesn't need to be involved. Other programs aren't something to be afraid of. :-P

<hr>

> If I'd to implement that I'd go for FTP as it seems easier todo in Lua.

Implement what? It's already there for anyone to use. :-)

**Comment by mwoz123** [comment]
> Hm? syncthing/syncthing-android#843

Today installed Syncthing: I opened chrome, open a webpage ->  "Share via" -> save in Syncthing -> error : "Nothing to share"

>You would just run Syncthing. KOReader doesn't need to be involved. Other programs aren't something to be afraid of. :-P

Right, as long as you're reading on Android E-ink (I do), but Kindle, Kobo, Pocketbook,etc readers would be disappointed... As there no multitasking in my Android, and to switching to other app it's too much taps (fro Koreader open menu-> hamburger-> exit -> exit -> enable wifi -> open anther app-> sync, quit, disable wifi -> run koreader, go to dir, etc) (I'm lazy;P) I'd prefer to have it integrated with Koreader.

>If I'd to implement that I'd go for FTP as it seems easier todo in Lua.
>Implement what? It's already there for anyone to use. :-)

Implement button (menu) that will login into specified service (FTP?), go to Article folder, copy files to specified folder on device, remove copied files from server.

**Comment by Frenzie** [comment]
> Right, as long as you're reading on Android E-ink (I do), but Kindle, Kobo, Pocketbook,etc readers would be disappointed...

That's not a particularly big issue. :-) It's fairly trivial on those platforms because they're not Android. But Android can nevertheless already do this today with greater ease *today* because the Syncthing people already solved all problems in the Syncthing app.

I'll quote what I outlined to @KenMaltby a while ago

> It could definitely be useful to do something with rsync, although I personally only use Unison and Dropbox/Syncthing (but not on my ereader, I mostly use wires for that).
>
> I'm inclined to think something like http://www.pebra.net/blog/2015/08/23/syncthing/ might be easier.
> 
> * Download syncthing for the platform (easy, at least on our Kobos)
> * Extract (easy, we already include tar)
> * Change the config for the web GUI
> * Display a little message to the user about where to point the browser

<hr>

> As there no multitasking in my Android, and to switching to other app it's too much taps (fro Koreader open menu-> hamburger-> exit -> exit -> enable wifi -> open anther app-> sync, quit, disable wifi -> run koreader, go to dir, etc) (I'm lazy;P) I'd prefer to have it integrated with Koreader.

I assume it can do multitasking by launching one app from another? For example, launching Syncthing from a KOReader menu entry, except that's not currently a thing. Because the key here is that we know the goal (reducing the number of actions). Refer to [how to ask a question the smart way](http://catb.org/~esr/faqs/smart-questions.html#goal). It's a bit… blunt in calling things stupid, which sounds offensive, but it is by and large correct in that the things it calls stupid are suboptimal as a mean to finding solutions.

> If you are trying to find out how to do something (as opposed to reporting a bug), begin by describing the goal. Only then describe the particular step towards it that you are blocked on.
>
> Often, people who need technical help have a high-level goal in mind and get stuck on what they think is one particular path towards the goal. They come for help with the step, but don't realize that the path is wrong. It can take substantial effort to get past this.
>
> **Stupid:**
>
>>    How do I get the color-picker on the FooDraw program to take a hexadecimal RGB value?
>
> **Smart:**
>
>>    I'm trying to replace the color table on an image with values of my choosing. Right now the only way I can see to do this is by editing each table slot, but I can't get FooDraw's color picker to take a hexadecimal RGB value.
>
> The second version of the question is smart. It allows an answer that suggests a tool better suited to the task.

It seems to me that a better solution toward the goal is to launch Syncthing, not to integrate Syncthing.

> Implement button (menu) that will login into specified service (FTP?), go to Article folder, copy files to specified folder on device, remove copied files from server.

That sounds fairly easy once you've got it setup, I suppose. :-) Personally I think Syncthing is easier to set up, while something like wallabag is probably not easier but more complete. But if you want to write an FTP folder download plugin (with optional deletion of downloaded items) that's really up to you. Also pinging @robert00s

**Comment by KenMaltby** [comment]
I think there is also the issue of Android App "Sharing" options, for instance my Fox News App has two pages of sharing options, including a WiFi to WiFi option (we would need the option to have our devices Discoverable, so it would be seen in the app).  This is on my Galaxy 7 Active

**Comment by mwoz123** [comment]
I did first initial coding for this feature in android. 

Current basic functionality :
- use "share via" to export url
- "downloads" url, 
- removes necessary tags from html (jsoup cleaner), reformat to valid html,  
- transform to epub (currently without images) 
- upload backend is interface (current only implementation is ftp folder, in future might be syncthing or some else as you want) 

As the above (except 1 point) is written in plain Java it is possible to extract that from android code to separate project (lib) so can be used from e.g. PC (as standalone Java app). 

It might be a good starting point for this feature. 

Do you want to add it to koreader ecosystem (new koreader related project) or you're not interested and I should keep it for myself?

**Comment by Frenzie** [comment]
My only worry is that the luajit-launcher attempts to be generic while this is fairly KOReader specific. But as far as functionality goes, sounds good! ;-)

**Comment by mwoz123** [comment]
> My only worry is that the luajit-launcher attempts to be generic while this is fairly KOReader specific


 As long as you have a tool to download from ftp and view epub (except the current work name ("sendToKoreader"))  it seems to be generic;)

as mainly all it does is generate epub and upload to ftp so nothing "vendor restricted"  to koreader only.

**Comment by Frenzie** [comment]
KOReader specific, ereader specific, potato, potato. :-P

I mean that the luajit-launcher is a generic way of making LuaJIT-based Android apps. All the functionality in there is present because we need it, but so far there's nothing unique to our kind of app.

**Comment by mwoz123** [comment]
Lol:) right it is eReader specific, I don't plan to build space ship from it :-Ppp

**Comment by mwoz123** [comment]
@Frenzie will you create a new repo on koreader I can push my code?

**Comment by Frenzie** [comment]
@mwoz123 Sure thing. Do you have a particular name in mind? We should probably prefix it with android- for clarity.

A couple of brainstorms:

android-share-webpage
android-webpagetoepub/webpage2epub

**Comment by poire-z** [comment]
android-ShareAsEPUB ?

> upload backend is interface (current only implementation is ftp folder, in future might be syncthing or some else as you want)

There can be a Koreader running on this same Android, and it would be nice if there was an alternative to the ftp/cloud destination: the local filesystem :)
So one could switch from his web browser to KOReader to continue reading with a far better reading app :)

**Comment by Frenzie** [comment]
I like that. I created https://github.com/koreader/android-shareasepub with an AGPLv3 license just like the main program but obviously we can still change it if desired. Let me know what you think.

**Comment by mwoz123** [comment]
Ehm name?  not yet. 
My workname was something like "web2ebook2cloud", "web2epub2cloud", "url2cloud", "sendToKoreader" 

I like the last one because it directly points at koreader so should be easy to find /Google. 
But it should could be used by other software, so might be a little bit confusing. 

Anyway can be changed later on, once we find better name. 

@poire-z hdd backend shouldn't be hard as it based on input /output streams. 


I still got couple refactoring in mind if I found time tomorrow/next days I'll do them, after that I'll push it. 

@KenMaltby would c
command line version something like :
> java sendToKoreader www.example.com www.otherurl.eu ....

will it be useful for you?  

It's still very early phase (especially android part ). Except that couple lines I wrote for koreader it's my first android code. So please get rid of any expectations here (actual version had nearly no gui, credentials read from property file) :p

I don't care about the license, as long it's compatible with 3rd party (jsoup,  https://github.com/psiegman/epublib/blob/master/CREDITS )

**Comment by mwoz123** [comment]
"sendToEreader/sendToReader"? As poire-z wants saving to disk "sendToCloud" doesn't seems to be good choice

**Comment by poire-z** [comment]
( It's not really that I want it - I don't use my phone for much more than phone calls :)
I was just explaining why **Share As Epub** could be a good name.

I don't know much about Android, but from the bits I read about, I would expect your app to work like this:
- In web browser => Share => type/intent = `"url"`
- Android would display all the apps that says they can handle `"url"`, including your **Share As Epub** app.
- When I chose **Share As Epub**, your app gets the url, and then does its real work: fetching the url, generating an EPUB file.
- Once that is done, your app would send an Intent `"mimetype=epub"`.
- Android would then list all the _other_ apps that say they can handle `"mimetype=epub"`, including KOReader, CoolReader, FBReader... (but I don't know how you would give them the epub... get the path to where to save the EPUB to? which should be a common part of the filesystem that all apps can access).
- And I hope you can there add all the remote backends of your choice (FTP...), may be just by having **Share As Epub** also announces that it can handle `"mimetype=epub"`, so another activity(?) of your app would get the epub and deal with uploading it somewhere ?

**Comment by mwoz123** [comment]
@poire-z  that's interesting idea. 
But there's still a lot more urget work:/ 

BTW I didn't want to use "Epub" in name as it can be also easily interfaced with other formats.

I pushed initial commit for app Java core:
https://github.com/mwoz123/send2ebook

To give you idea how it works view main():
https://github.com/mwoz123/send2ebook/blob/master/src/main/java/com/github/mwoz123/send2ebook/Send2Ebook.java

It's still early release but should be usable from console 
(after creating connection Properties, [sample one here](https://github.com/mwoz123/send2ebook/blob/master/src/main/resources/sampleConnection.properties)  and adding parameter `-Dconnection.file.path=/path_to_your_connection_file.properties` to commandline )

**Comment by Frenzie** [comment]
@mwoz123 Did you receive the collaborator invitation e-mail for the repo I made a couple of days ago? Either way, you should be able to do it from https://github.com/koreader/android-send2ebook/invitations

**Comment by mwoz123** [comment]
Push init draft of Send2Ebook-android to:
https://github.com/koreader/android-send2ebook

Unfortunately after last refactor - exacting [java part](https://github.com/mwoz123/send2ebook) and importing it as lib, I did something wrong as I've error:
```
Error:Error converting bytecode to dex:
Cause: Dex cannot parse version 53 byte code.
```
Couple top stackoverflow solutions didn't work for me, so if someone could take a look, would be great.

**Comment by Frenzie** [comment]
I opened it in Android Studio and got this. Not sure if that's related to what you're seeing or just something wrong in my config.

> Warning:The `android.dexOptions.incremental` property is deprecated and it has no effect on the build process.

> Error:Execution failed for task ':app:javaPreCompileDebug'.
> Annotation processors must be explicitly declared now.  The following dependencies on the compile classpath are found to contain annotation processor.  Please add them to the annotationProcessor configuration.
    - send2ebook-with-dependecies-0.7.jar (send2ebook-with-dependecies-0.7.jar)
  Alternatively, set android.defaultConfig.javaCompileOptions.annotationProcessorOptions.includeCompileClasspath = true to continue with previous behavior.  Note that this option is deprecated and will be removed in the future.
  See https://developer.android.com/r/tools/annotation-processor-error-message.html for more details.

![screenshot_2018-02-10_20-27-15](https://user-images.githubusercontent.com/202757/36065847-ce2f3818-0ea0-11e8-95a9-6b250a5335fa.png)

**Comment by mwoz123** [comment]
I recreated from scratch android part with latest Android Studio.
It's in https://github.com/koreader/android-send2ebook/tree/new-start branch.

@Frenzie Could you remove 'master', and rename "new-start" to 'master'? 

Should be quite usable, + got UI configuration;)

**Comment by mwoz123** [comment]
Added downloading images. Build apk can be found here:
https://github.com/koreader/android-send2ebook/tree/new-start/repo

@Frenzie can we change license to less virus-like? I like the LGPL. 
Can you replace 'master' branch with 'new-start' or should I do push force?

**Comment by Frenzie** [comment]
LGPL should be fine.

You normally shouldn't put binaries in a Git repo. You can sneak them into GitHub comments btw, either stick .zip at the end (which for apk is rather appropriate) or package it inside a zip file. :-)

Via "create release" you can also upload binaries attached to a specific tag.

Looks like you can't rename branches through GH, but in this case it sounds like force pushing the new-start branch to master is really all that's desired?

**Comment by silberzwiebel** [comment]
Oh, that sounds like an interesting new feature, happy to test once it's merged!

Just adding my two cents (as I'm an almost happy user of wallabag on my android e-ink device):
Another not-yet-mentioned pro of wallabag is it's open sourceness, which allows you to deploy your own server (which isn't that hard) and thus you can keep control of your (meta-)data on what you find interesting to read (Plus: wallabag allows organizing articles with tags, favorites, marks as read).

I didn't read through all details, but with the current proposal for KOreader's send2ebook there is some FTP server involved? Depending on where this server is located, I'd see privacy problems here (although of course no-one is forced to use the plugin).

Second question: Would this work from a PC, e.g. as a firefox plugin?

**Comment by Frenzie** [comment]
> Another not-yet-mentioned pro of wallabag is it's open sourceness

Possibly not explicitly mentioned, but that's the reason I brought up Syncthing and wallabag specifically rather than, say, Dropbox and Pocket.

> I didn't read through all details, but with the current proposal for KOreader's send2ebook there is some FTP server involved? Depending on where this server is located, I'd see privacy problems here (although of course no-one is forced to use the plugin).

There are open-source FTP servers galore, so I'm not quite sure just what you're getting at. :-P

**Comment by silberzwiebel** [comment]
> Possibly not explicitly mentioned, but that's the reason I brought up Syncthing and wallabag specifically rather than, say, Dropbox and Pocket.

Pocket should be open-sourced but apparently they are not fast in doing so. Only the client code is open source yet, not the server code.

> There are open-source FTP servers galore, so I'm not quite sure just what you're getting at. :-P

Sure, I was just wondering if and how and where a FTP server needs to be deployed / is deployed by someone / ...

**Comment by mwoz123** [comment]
@silberzwiebel you don't need to create your own ftp server.
Just search internet for "(best) free hosting" or "free ftp hosting" there are really plenty of them.
One of the results :
https://www.techradar.com/news/internet/web/top-free-hosting-for-uk-businesses-1214296

Just find one popular nearby you location or at least not oversees;) (might be some faster)

About the privacy, as you're (current implementation) passing just a http link to Send2Ebook, it behaves as web browser, so if the page is not public (requires login), currently (as it should be redirect to login page) it will only create ebook with login page inside (not tested that scenario but that's what I assume).
Anyway As far as I know in ftp servers you restrict access to folders for selected users, so you can create folder where only you might have access to it.

There's a no plugin for PC firefox/chrome etc (it would have be written in Javascript or please let me know if there's adapter/converter to run Java from JS), instead you can use [commandline](https://github.com/mwoz123/send2ebook/tree/master/commandline) version (maybe someone could write PC (java?) GUI?) I know currently there's a bug but once fixed you should be able to use (unixes):
`send2ebook.sh www.a_url.com www.second_url.com` (..)
or (windows) :
`send2ebook.bat ww.a_url.com www.second_url` (..)

cheers

**Comment by Frenzie** [comment]
> if there's adapter/converter to run Java from JS

There's https://github.com/cincheo/jsweet though I doubt it'd be much use in this case. The only way to upload to an FTP server from JS afaik is through an HTTP intermediary.

**Comment by mwoz123** [comment]
> You normally shouldn't put binaries in a Git repo. You can sneak them into GitHub comments btw, either stick .zip at the end (which for apk is rather appropriate) or package it inside a zip file. :-)

the most benefits I see is I can download and install apk directly from my phone.
That's why I prefer apk.
Is there a chance to add it to koreader's CI?

**Comment by Frenzie** [comment]
> the most benefits I see is I can download and install apk directly from my phone.

But besides not belonging in the repo in the first place, you'd have to jump through hoops like "raw file" instead of the nice convenient purpose-made release tab. >_>

> Is there a chance to add it to koreader's CI?

Sure, just add some Travis or CircleCI config. Only on tags, of course. CircleCI is easier to debug but since I doubt we're looking at more than a few minutes either way I doubt it matters much.

https://docs.travis-ci.com/user/deployment/releases/
https://jedkirby.com/blog/github-releases-with-circleci

**Comment by poire-z** [comment]
This has been implemented in #3681.
